File: 프로그래머스/1/172928. 공원 산책/공원 산책.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<string> park, vector<string> routes) {
    int x, y;
    for(int i=0; i<park.size(); i++) {
        for(int j=0; j<park[i].size(); j++) {
            if(park[i][j] == 'S') x = i, y = j;
        }
    }
    vector<int> answer;
    for(int i=0; i<routes.size(); i++) {
        if(routes[i][0] == 'N') {
            bool check = true;
            int cnt = routes[i][2] - '0';
            if(x - cnt >= 0) {
                for(int j=1; j<=cnt; j++) {
                    if(park[x-j][y] == 'X') check = false;
                }
                if(check) x = x - cnt, y = y;
            }
            else check = false;
        }
        if(routes[i][0] == 'S') {
            bool check = true;
            int cnt = routes[i][2] - '0';
            if(x + cnt < park.size()) {
                for(int j=1; j<=cnt; j++) {
                    if(park[x+j][y] == 'X') check = false;
                }
                if(check) x = x + cnt, y = y;
            }
        }
        if(routes[i][0] == 'W') {
            bool check = true;
            int cnt = routes[i][2] - '0';
            if(y - cnt >= 0) {
                for(int j=1; j<=cnt; j++) {
                    if(park[x][y-j] == 'X') check = false;
                }
                if(check) x = x, y = y - cnt;
            }
        }
        if(routes[i][0] == 'E') {
            bool check = true;
            int cnt = routes[i][2] - '0';
            if(y + cnt < park[0].size()) {
                for(int j=1; j<=cnt; j++) {
                    if(park[x][y+j] == 'X') check = false;
                }
                if(check) x = x, y = y + cnt;
            }
        }
    }
    answer.push_back(x);
    answer.push_back(y);
    return answer;
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.swift
import Foundation

func solution(_ park:[String], _ routes:[String]) -> [Int] {
    let col: Int = park.count
    let row: Int = park[0].length
    var map: [[Int]] = Array(repeating: Array(repeating: 0, count: 50), count : 50)
    var curr: [Int] = [0, 0]
    
    for i in 0 ..< park.count {
        for j in 0 ..< park[i].length {
            let curIdx = park[i].index(park[i].startIndex, offsetBy: j)
            if String(park[i][curIdx]) == "S" {
                curr[0] = i
                curr[1] = j
            } else if String(park[i][curIdx]) == "X" {
                map[i][j] = 1
            }
        }
    }
    
    for route in routes {
        let dir = route[route.index(route.startIndex, offsetBy: 0)]
        let cnt: Int = Int(String(route[route.index(route.startIndex, offsetBy: 2)]))!
        
        var check: Bool = false
        
        if dir == "E" && curr[1] + cnt < row {
            for i in 1...cnt {
                if map[curr[0]][curr[1] + i] == 1 {
                    check = true;
                }
            }
            if !check {
                curr[1] += cnt
            }
        } else if dir == "W" && curr[1] - cnt >= 0 {
            for i in 1...cnt {
                if map[curr[0]][curr[1] - i] == 1 {
                    check = true;
                }
            }
            if !check {
                curr[1] -= cnt
            }
        } else if dir == "N" && curr[0] - cnt >= 0 {
            for i in 1...cnt {
                if map[curr[0] - i][curr[1]] == 1 {
                    check = true;
                }
            }
            if !check {
                curr[0] -= cnt
            }
        } else if dir == "S" && curr[0] + cnt < col {
            for i in 1...cnt {
                if map[curr[0] + i][curr[1]] == 1 {
                    check = true;
                }
            }
            if !check {
                curr[0] += cnt
            }
        }
    }
    
    return curr
}